mat64/list_test.go

```diff
@@ -0,0 +1,706 @@
+			if a.At(i, j) != b.At(i, j) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// testTwoInput tests a method that has two input arguments.
+// receiver is a value of the receiver type.
+// method is receiver.Method(a,b). Returns false if the matrix types aren't correct.
+// denseComparison performs the same operation but with dense matrices for comparison
+// with the result.
+// legalTypes returns false if the concrete types in Matrix are compile-time
+// provably invalid for this method.
+// legalSize returns false if the matrix sizes aren't okay for the function.
+func testTwoInput(c *check.C,
+	receiver Matrix,
+	method func(receiver, a, b Matrix),
+	denseComparison func(receiver, a, b *Dense),
+	legalTypes func(a, b Matrix) bool,
+	legalSize func(ar, ac, br, bc int) bool,
+) {
+	strideVec := &Vector{
+		mat: blas64.Vector{
+			Inc: 10,
+		},
+	}
+	// Loop over all of the matrix types.
+	matrices := []Matrix{
+		&Dense{},
+		&SymDense{},
+		NewTriDense(0, true, nil),
+		NewTriDense(0, false, nil),
+		NewVector(0, nil),
+		strideVec,
+		&basicMatrix{},
+		&basicVectorer{},
+		&basicSymmetric{},
+		&basicTriangular{},
+	}
+
+	for _, aMat := range matrices {
+		for _, bMat := range matrices {
+			// Loop over all of the size combinations (bigger, smaller, etc.)
+			for _, test := range []struct {
+				ar, ac, br, bc int
+			}{
+
+				{1, 1, 1, 1},
+				{6, 6, 6, 6},
+				{7, 7, 7, 7},
+
+				{1, 1, 1, 5},
+				{1, 1, 5, 1},
+				{1, 5, 1, 1},
+				{5, 1, 1, 1},
+				{6, 6, 6, 11},
+				{6, 6, 11, 6},
+				{6, 11, 6, 6},
+				{11, 6, 6, 6},
+				{11, 11, 11, 6},
+				{11, 11, 6, 11},
+				{11, 6, 11, 11},
+				{6, 11, 11, 11},
+
+				{1, 1, 5, 5},
+				{1, 5, 1, 5},
+				{1, 5, 5, 1},
+				{5, 1, 1, 5},
+				{5, 1, 5, 1},
+				{5, 5, 1, 1},
+				{6, 6, 11, 11},
+				{6, 11, 6, 11},
+				{6, 11, 11, 6},
+				{11, 6, 6, 11},
+				{11, 6, 11, 6},
+				{11, 11, 6, 6},
+
+				{1, 1, 17, 11},
+				{1, 1, 11, 17},
+				{1, 11, 1, 17},
+				{1, 17, 1, 11},
+				{1, 11, 17, 1},
+				{1, 17, 11, 1},
+				{11, 1, 1, 17},
+				{17, 1, 1, 11},
+				{11, 1, 17, 1},
+				{17, 1, 11, 1},
+				{11, 17, 1, 1},
+				{17, 11, 1, 1},
+
+				{6, 6, 1, 11},
+				{6, 6, 11, 1},
+				{6, 11, 6, 1},
+				{6, 1, 6, 11},
+				{6, 11, 1, 6},
+				{6, 1, 11, 6},
+				{11, 6, 6, 1},
+				{1, 6, 6, 11},
+				{11, 6, 1, 6},
+				{1, 6, 11, 6},
+				{11, 1, 6, 6},
+				{1, 11, 6, 6},
+
+				{6, 6, 17, 1},
+				{6, 6, 1, 17},
+				{6, 1, 6, 17},
+				{6, 17, 6, 1},
+				{6, 1, 17, 6},
+				{6, 17, 1, 6},
+				{1, 6, 6, 17},
+				{17, 6, 6, 1},
+				{1, 6, 17, 6},
+				{17, 6, 1, 6},
+				{1, 17, 6, 6},
+				{17, 1, 6, 6},
+
+				{6, 6, 17, 11},
+				{6, 6, 11, 17},
+				{6, 11, 6, 17},
+				{6, 17, 6, 11},
+				{6, 11, 17, 6},
+				{6, 17, 11, 6},
+				{11, 6, 6, 17},
+				{17, 6, 6, 11},
+				{11, 6, 17, 6},
+				{17, 6, 11, 6},
+				{11, 17, 6, 6},
+				{17, 11, 6, 6},
+			} {
+				// Skip the test if the arguments are compile-time provably incorrect
```

Comment:
The comment was talking about the call
````
sym.Mul(sym,dense)
````
This could never happen in user code because Dense does not implement Symmetric. It's compile time impossible for those matrix types to be used. Similarly, the type/size loops above could call for a 17x6 symmetric matrix. This is impossible; the symmetric type must be square.